File: src/sc_neurocore/accel/julia/layers/recurrent.jl
# SPDX-License-Identifier: AGPL-3.0-or-later
# Commercial license available
# © Concepts 1996–2026 Miroslav Šotek. All rights reserved.
# © Code 2020–2026 Miroslav Šotek. All rights reserved.
# ORCID: 0009-0009-3560-0851
# Contact: www.anulum.li | protoscience@anulum.li
# SC-NeuroCore — Julia acceleration for layers/recurrent

module RecurrentAccel

using Statistics, LinearAlgebra

mutable struct SCRecurrentLayerState
    n_inputs::Float64
    n_neurons::Float64
    feedback_strength::Float64
    input_strength::Float64
    spectral_radius::Float64
    length::Float64
    seed::Float64
end

function SCRecurrentLayerState()
    SCRecurrentLayerState(0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0)
end

function step(s::SCRecurrentLayerState, input_vector, Any])
    currents = dot(s.W_in, input_vector) + dot(s.W_rec, s.state)
    new_rates = clamp(currents, 0.0, 1.0)
    s.state = new_rates
    return s.state
end

function reset(s::SCRecurrentLayerState)
    s.state = zeros(s.n_neurons)
end

end # module RecurrentAccel

File: src/sc_neurocore/accel/rust/safety/recurrent.rs
// SPDX-License-Identifier: AGPL-3.0-or-later
// Commercial license available
// © Concepts 1996–2026 Miroslav Šotek. All rights reserved.
// © Code 2020–2026 Miroslav Šotek. All rights reserved.
// ORCID: 0009-0009-3560-0851
// Contact: www.anulum.li | protoscience@anulum.li
// SC-NeuroCore — Rust safety for recurrent

#![allow(unused_variables, dead_code, non_snake_case)]

#[derive(Debug, Clone)]
pub struct SCRecurrentLayer {
    pub n_inputs: f64,
    pub n_neurons: f64,
    pub feedback_strength: f64,
    pub input_strength: f64,
    pub spectral_radius: f64,
    pub length: f64,
    pub seed: f64,
}

impl SCRecurrentLayer {
    pub fn new() -> Self {
        Self {
            n_inputs: 0.0_f64,
            n_neurons: 0.0_f64,
            feedback_strength: 0.0_f64,
            input_strength: 0.0_f64,
            spectral_radius: 0.0_f64,
            length: 0.0_f64,
            seed: 0.0_f64,
        }
    }

    pub fn step(&mut self, i_ext: f64) -> i32 {
        // currents = np.dot(self.W_in, input_vector) + np.dot(self.W_rec, self.s
        // new_rates = (currents_f64).clamp(0.0, 1.0)
        // self.state = new_rates
        // return self.state
        0 // spike indicator
    }

    pub fn reset(&mut self) {
        // self.state = np.zeros(self.n_neurons)
        self.n_inputs = 0.0_f64;
        self.n_neurons = 0.0_f64;
        self.feedback_strength = 0.0_f64;
        self.input_strength = 0.0_f64;
        self.spectral_radius = 0.0_f64;
    }

}

pub fn validate_recurrent(state: &SCRecurrentLayer) -> bool {
    true
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_recurrent_new() {
        let state = SCRecurrentLayer::new();
        assert!(validate_recurrent(&state));
    }

    #[test]
    fn test_recurrent_step() {
        let mut state = SCRecurrentLayer::new();
        let spike = state.step(10.0);
        assert!(spike == 0 || spike == 1);
    }
}
